tests/06_REST_oauth2.js

```diff
@@ -0,0 +1,128 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * Except as contained in this notice, the name of OVH and or its trademarks
+ * shall not be used in advertising or otherwise to promote the sale, use or
+ * other dealings in this Software without prior written authorization from OVH.
+ */
+
+var fs = require('fs'),
+    ovh = require('../lib/ovh.es5'),
+    assert = require('assert'),
+    async = require('async'),
+    nock = require('nock');
+
+var oauthConfig = {
+  clientID: '1234',
+  clientSecret: 'abcdef',
+  endpoint: 'ovh-eu'
+};
+
+exports.REST_oauth2 = {
+  'GET /me with a valid token': function (done) {
+    'use strict';
+
+    nock('https://www.ovh.com/auth')
+     .intercept('/oauth2/token', 'POST')
+       .reply(200, {
+          'access_token': 'ok',
+          'token_type': 'Bearer',
+          'expires_in': 3599,
+          'scope': 'all'
+       });
+    nock('https://eu.api.ovh.com')
+     .intercept('/1.0/me', 'GET')
+       .reply(200, {"status": "ok"});
+
+    var rest = ovh(oauthConfig);
+    rest.request('GET', '/me', function (err, msg) {
+      assert.ok(!err);
+      assert.equal(msg.status, 'ok');
+      done();
+    });
+  },
+  'GET /me with a valid token [promised]': function (done) {
+    'use strict';
+
+    nock('https://www.ovh.com/auth')
+     .intercept('/oauth2/token', 'POST')
+       .reply(200, {
+          'access_token': 'ok',
+          'token_type': 'Bearer',
+          'expires_in': 3599,
+          'scope': 'all'
+       });
+    nock('https://eu.api.ovh.com')
+     .intercept('/1.0/me', 'GET')
+       .reply(200, {"status": "ok"});
+
+    var rest = ovh(oauthConfig);
+    rest.requestPromised('GET', '/me')
+      .then(function (msg) {
+        assert.equal(msg.status, 'ok');
+      })
+      .catch(function (err) {
+        assert.ok(!err);
+      })
+      .finally(done);
+  },
+  'GET /me with invalid credentials': function (done) {
+    'use strict';
+
+    nock('https://www.ovh.com/auth')
+     .intercept('/oauth2/token', 'POST')
+       .reply(401, {
+          'error': 'Invalid client'
+       });
+
+    var rest = ovh(oauthConfig);
+    rest.request('GET', '/me', function (err, msg) {
+      assert.equal(err, `[OVH] failed to retrieveOAuth2 Access Token`);
+      done();
+    });
+  },
+  'GET /me with expired token': function (done) {
+    'use strict';
+
+    nock('https://www.ovh.com/auth')
+     .intercept('/oauth2/token', 'POST')
+     .reply(200, {
+      'access_token': 'ok',
+      'token_type': 'Bearer',
+      'expires_in': 3599,
+      'scope': 'all'
+   });
+   nock('https://eu.api.ovh.com')
+   .intercept('/1.0/me', 'GET')
+     .reply(200, {"status": "ok"});
+
+    var rest = ovh(oauthConfig);
+    rest.accessToken = {
+      expired: function() { return true; }
```

Comment:
I'm not sure of what this is testing, expired or not, a unique call will always fetch a token.
I would write something like that, in my own classy JS style OFC

```js
'GET /me with expired token': function (done) {
  'use strict';

  this.timeout(5000);

  // First call, expires soon
  nock('https://www.ovh.com/auth')
    .post('/oauth2/token')
    .reply(200, {
      'access_token': 'TOKEN1',
      'token_type': 'Bearer',
      'expires_in': 10,
      'scope': 'all'
    });

  // Refresh call
  nock('https://www.ovh.com/auth')
    .post('/oauth2/token')
    .reply(200, {
      'access_token': 'TOKEN2',
      'token_type': 'Bearer',
      'expires_in': 3599,
      'scope': 'all'
    });

  // Called once with TOKEN1
  nock('https://eu.api.ovh.com', {
    reqheaders: {
      authorization: 'Bearer TOKEN1',
    },
  })
    .get("/1.0/me")
    .reply(200, {"status": "ok1"});

  // Called again with TOKEN2, after the refresh
  nock('https://eu.api.ovh.com', {
    reqheaders: {
      authorization: 'Bearer TOKEN2',
    },
  })
    .get("/1.0/me")
    .reply(200, {"status": "ok2"});

  var rest = ovh(oauthConfig);

  rest.request('GET', '/me', function (err, msg) {
    assert.ok(!err);
    assert.equal(msg.status, 'ok1');

    const sleep = ms => new Promise(r => setTimeout(r, ms));
    sleep(2000);

    rest.request('GET', '/me', function (err, msg) {
      assert.ok(!err);
      assert.equal(msg.status, 'ok2');

      done();
    });
  });
},

```